### PR TITLE
Add metric telemetry 'MarkdownElement'

### PR DESCRIPTION
**Pull request by 928PJY**
[AB#193621](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/193621)
Design Spec: https://review.docs.microsoft.com/en-us/engineering/projects/ops/build/telemetry/build-telemetry?branch=master#markdown-element

And enable telemetry in integration test to improve test coverage

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5719)

### CODE REVIEW

**Comment by mingwli** [comment]
> ⚠[mc-docs-pr](https://github.com/opse2epr/mc-docs-pr)(00:01:03.3816774, 5526 more diff)

I'm creating a dedicated branch. Orzz

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 1]
```diff
@@ -13,6 +17,34 @@ namespace Microsoft.Docs.Build
 {
     internal static class MarkdigUtility
     {
+        private static readonly IReadOnlyDictionary<Type, string> s_markdownElementTypeMapping = new Dictionary<Type, string>()
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:20]: I tent to like putting this into `MarkdownElementTelemetryExtension`, the `MarkdigUtility` was intended for common shared functions, this function is telemetry only, so

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 2]
```diff
@@ -112,6 +144,102 @@ public static MarkdownPipelineBuilder Use(this MarkdownPipelineBuilder builder,
             return builder;
         }
 
+        public static string? GetElementType(MarkdownObject node)
+        {
+            if (node is null)
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:149]: with nullable checks turned on, there is no need to check `null` here as the parameter type is marked as `non-nullable`

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 3]
```diff
@@ -112,6 +144,102 @@ public static MarkdownPipelineBuilder Use(this MarkdownPipelineBuilder builder,
...
+            {
+                return default;
+            }
+
+            string? elementType = s_markdownElementTypeMapping.GetValueOrDefault(node.GetType(), null);
+            if (elementType is null)
+            {
+                if (node is HeadingBlock heading)
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:157]: Is this a perfect place for [switch expression](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#switch-expressions)?

File: src/docfx/lib/log/Telemetry.cs [hunk 4]
```diff
@@ -120,6 +122,17 @@ public void Dispose()
...
+        private static (string fileExtension, string documentType, string mimeType) GetFileType(Document file)
+        {
+            var fileExtension = CoalesceEmpty(Path.GetExtension(file.FilePath.Path)?.ToLowerInvariant());
+            var mimeType = CoalesceEmpty(file.Mime.Value);
+            if (mimeType == "<null>" && file.ContentType == ContentType.Page && fileExtension == ".md")
+            {
+                mimeType = "Conceptual";
+            }
+            return (fileExtension, file.ContentType.ToString(), mimeType);
+        }
```
Comment (mingwli) [inline src/docfx/lib/log/Telemetry.cs:134]: need this be cached?

Comment (928PJY) [reply src/docfx/lib/log/Telemetry.cs:134]: good point, this function could be called plenty of times for the same file

---
In reply to: [402153515](https://github.com/dotnet/docfx/pull/5719#discussion_r402153515) [](ancestors = 402153515)

**Comment by mingwli** [comment]
> 🧭[azure-docs-pr](https://github.com/OPS-E2E-PPE/azure-docs-pr)(00:04:28.9769362 | exceed 150s)

Caused by first time run on the new agent, restore took longer.

File: src/docfx/lib/markdown/MarkdownElementTelemetryExtension.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                ParagraphBlock paragraphBlock => "Paragraph",
+                BlankLineBlock blankLineBlock => "BlankLine",
+                QuoteSectionNoteBlock quoteSectionNoteBlock =>
+                    quoteSectionNoteBlock.QuoteType switch
+                    {
+                        QuoteSectionNoteType.DFMNote => "Note",
+                        QuoteSectionNoteType.DFMSection => "SectionDefinition",
+                        QuoteSectionNoteType.DFMVideo => "Video",
+                        QuoteSectionNoteType.MarkdownQuote => "BlockQuote",
+                        _ => throw new NotSupportedException(),
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdownElementTelemetryExtension.cs:51]: better not throw here, line 76 return null for unknown element, same here?

File: src/docfx/lib/markdown/MarkdownElementTelemetryExtension.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                HtmlBlock htmlBlock => $"-{htmlBlock.Type}",
+                QuoteSectionNoteBlock quoteSectionNoteBlock => GetQuoteSectionNoteBlockSubType(quoteSectionNoteBlock),
+                TripleColonBlock tripleColonBlock => $"-{StringUtility.UpperCaseFirstChar(tripleColonBlock.Extension.Name)}",
+                _ => null,
+            };
+
+            return tokenType;
+
+            string GetQuoteSectionNoteBlockSubType(QuoteSectionNoteBlock quoteSectionNote)
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdownElementTelemetryExtension.cs:93]: nit: this could also be an expression

**Pull request merged**